wallet-new/src/Cardano/Wallet/Kernel/Addresses.hs

```diff
@@ -0,0 +1,158 @@
+         -- 'HdAddress' we need a proper 'Address', but this cannot be derived
+         -- with having access to the 'EncryptedSecretKey' and the 'PassPhrase',
+         -- and we do not want these exposed in the acid-state transaction log.
+         (AccountIdHdRnd hdAccId) -> do
+             mbEsk <- Keystore.lookup (WalletIdHdRnd (hdAccId ^. hdAccountIdParent))
+                                      keystore
+             case mbEsk of
+                  Nothing  -> return (Left $ CreateAddressKeystoreNotFound accId)
+                  Just esk -> createHdRndAddress spendingPassword esk hdAccId pw
+
+-- | Creates a new 'Address' using the random HD derivation under the hood.
+-- Being this an operation bound not only by the number of available derivation
+-- indexes \"left\" in the account, some form of short-circuiting is necessary.
+-- Currently, the algorithm is as follows:
+--
+-- 1. Try to generate an 'Address' by picking a random index;
+-- 2. If the operation succeeds, return the 'Address';
+-- 3. If the DB operation fails due to a collision, try again, up to a max of
+--    2147483647 attempts (which is half of the address space).
+-- 4. If after 2147483647 attempts there is still no result, flag this upstream.
+createHdRndAddress :: PassPhrase
+                   -> EncryptedSecretKey
+                   -> HdAccountId
+                   -> PassiveWallet
+                   -> IO (Either CreateAddressError Address)
+createHdRndAddress spendingPassword esk accId pw = do
+    gen <- createSystemRandom
+    go gen 0
+    where
+        go :: GenIO -> Word32 -> IO (Either CreateAddressError Address)
+        go gen collisions =
+            case collisions >= maxAllowedCollisions of
+                 True  -> return $ Left (CreateAddressHdRndAddressSpaceSaturated accId)
+                 False -> tryGenerateAddress gen collisions
+
+        tryGenerateAddress :: GenIO
+                           -> Word32
+                           -- ^ The current number of collisions
+                           -> IO (Either CreateAddressError Address)
+        tryGenerateAddress gen collisions = do
+            newIndex <- deriveIndex (flip uniformR gen) HdAddressIx HardDerivation
+            let hdAddressId = HdAddressId accId newIndex
+                mbAddr = deriveLvl2KeyPair (IsBootstrapEraAddr True)
+                                           (ShouldCheckPassphrase True)
+                                           spendingPassword
+                                           esk
+                                           (accId ^. hdAccountIdIx . to getHdAccountIx)
+                                           (hdAddressId ^. hdAddressIdIx . to getHdAddressIx)
+            case mbAddr of
+                 Nothing -> return (Left $ CreateAddressHdRndGenerationFailed accId)
+                 Just (newAddress, _) -> do
+                    let hdAddress  = initHdAddress hdAddressId (InDb newAddress)
```

Comment:
NOTE: We clarified with @edsko that is actually fine to _not_ worry about `isUsed` and `isChange` as that were a relic of an old model that the spec (with its BlockMeta) superseded, so @uroboros work will make sure we will remove them from the acid state model completely.